pkg/codegen/python/gen.go

```diff
@@ -513,8 +513,61 @@ def get_version():
      return _version_str
 `)
 	}
+	if err != nil {
+		return nil, err
+	}
+
+	if pkg.Parameterization != nil {
+		// If a parameterized package is being generated then we _need_ to use package references
+
+		buildValue := func(value interface{}) string {
+			switch v := value.(type) {
+			case string:
+				return fmt.Sprintf("struct_pb2.Value(string_value=%q)", v)
+			default:
+				panic(fmt.Sprintf("unsupported parameter value type %v", reflect.TypeOf(value)))
+			}
+		}
+
+		param := buildValue(pkg.Parameterization.Parameter)
+
+		_, err = fmt.Fprintf(buffer, `
+_package_lock = asyncio.Lock()
+_package_ref = ...
+async def get_package():
+	global _package_ref
+	if _package_ref is ...:
+		async with _package_lock:
+			if _package_ref is ...:
+				monitor = pulumi.runtime.settings.get_monitor()
+				parameter = resource_pb2.PackageParameter(
+					name=%q,
+					version=%q,
+					value=%s,
+				)
+				try:
+					registerPackageResponse = monitor.RegisterPackage(
+						resource_pb2.RegisterPackageRequest(
+							name=%q,
+							version=get_version(),
+							plugin_download_url=get_plugin_download_url(),
+							parameter=parameter,
+						))
```

Comment:
I'm a bit torn on this. Should the _plugin_ name always be the name and version in RegisterPackageRequest, and then the parameterised package name in PackageParameter. Or should it be like I've done here where the package name and version is in RegisterPackageRequest and the base plugin name and version are in PackageParameter.

I'm more and more inclined to switch this because the other plugin properties (like download url and checksums) are directly on RegisterPackageRequest. Feels like it would be better to keep all the plugin properties together. I even considered if this should be split into two sub-message "plugin" and "parameterisation" to make that even more explicit.

Reply:
I think I'm leaning towards this way as well, keeping the properties for the plugin together.  I think that's clearer.

Reply:
Yeh I'll raise a PR to flip the engine logic on this then rebase this PR.